qiskit_machine_learning/datasets/ad_hoc.py

```
@@ -84,140 +82,128 @@ def ad_hoc_data(
-        phi += sum(
-            ((np.pi - x_arr[i1]) * (np.pi - x_arr[i2]) * z_i[i1] @ z_i[i2] for i1, i2 in ind_pairs)
-        )
-        # u_u was actually scipy.linalg.expm(1j * phi), but this method is
-        # faster because phi is always a diagonal matrix.
-        # We first extract the diagonal elements, then do exponentiation, then
-        # construct a diagonal matrix from them.
-        u_u = np.diag(np.exp(1j * np.diag(phi)))
-        psi = u_u @ h_n @ u_u @ psi_0
-        exp_val = np.real(psi.conj().T @ m_m @ psi)
-        if np.abs(exp_val) > gap:
-            _sample_total.append(np.sign(exp_val))
-        else:
-            _sample_total.append(0)
-    sample_total = np.array(_sample_total).reshape(*[count] * n)
-
-    # Extract training and testing samples from grid
-    x_sample, y_sample = _sample_ad_hoc_data(sample_total, xvals, training_size + test_size, n)
-
-    if plot_data:
-        _plot_ad_hoc_data(x_sample, y_sample, training_size)
-
-    training_input = {
-        key: (x_sample[y_sample == k, :])[:training_size] for k, key in enumerate(class_labels)
-    }
-    test_input = {
-        key: (x_sample[y_sample == k, :])[training_size : (training_size + test_size)]
-        for k, key in enumerate(class_labels)
-    }
-
-    training_feature_array, training_label_array = _features_and_labels_transform(
-        training_input, class_labels, one_hot
-    )
-    test_feature_array, test_label_array = _features_and_labels_transform(
-        test_input, class_labels, one_hot
-    )
+    V = eigvecs[:, idx]
+
+    # Observable for labelling boundary
+    O = V.conj().T @ z_n @ V
+
+    # Loop for Data Acceptance & Regeneration
+    n_samples = training_size+test_size
+    features = np.empty((n_samples, n), dtype=float)
+    labels = np.empty(n_samples, dtype=int)
+    cur = 0
+
+    while n_samples > 0:
+        # Stratified Sampling for x vector
+        if divisions>0: x_vecs = _modified_LHC(n, n_samples, divisions)
```

In my view one way to incorporate the back-compatible
```python
basis = algorithm_globals.random.random(
        (2**n, 2**n)
    ) + 1j * algorithm_globals.random.random((2**n, 2**n))
```
would be to introduce another function, say `_uniform_havlicek`, that implements it and is triggered in an elif logic like this. To make the current unit tests pass, then you'd select this sampling type.

If you were to go ahead with this, then I'd suggest introducing a function argument called eg `sampling_method`, that accept the Literal values "sobol", "hypercube", or "havlicek", and if the use selects "hypercube", then they must pass `divisions` too.

Then, you could raise an error in `_uniform_havlicek` if n>3, specifying that this (original) method only works for 2 and 3, or you could try to extend it to arbitrary n in a way that it reduces to the original form when n=2, 3.